// @react-native-firebase/analytics/dist/commonjs/web/RNFBAnalyticsModule.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RNFBAnalyticsModule = void 0;
var _firebaseApp = require("@react-native-firebase/app/lib/internal/web/firebaseApp");
var _utils = require("@react-native-firebase/app/lib/internal/web/utils");
var _api = require("./api.js");
let analyticsInstances = {};
function getAnalyticsApi(appName) {
  const app = (0, _firebaseApp.getApp)(appName);
  const measurementId = app.options.measurementId;
  if (!measurementId) {
    // eslint-disable-next-line no-console
    console.warn('No measurement id (`FirebaseOptions.measurementId`) found for Firebase Analytics. Analytics will be unavailable.');
    // Return a default instance with empty measurementId for cases where it's not configured
    const defaultKey = 'default';
    if (!analyticsInstances[defaultKey]) {
      analyticsInstances[defaultKey] = new _api.AnalyticsApi('[DEFAULT]', '');
    }
    return analyticsInstances[defaultKey];
  }
  if (!analyticsInstances[measurementId]) {
    analyticsInstances[measurementId] = new _api.AnalyticsApi('[DEFAULT]', measurementId);
    if (globalThis.RNFBDebug) {
      analyticsInstances[measurementId].setDebug(true);
    }
  }
  return analyticsInstances[measurementId];
}

/**
 * This is a 'NativeModule' for the web platform.
 * Methods here are identical to the ones found in
 * the native android/ios modules e.g. `@ReactMethod` annotated
 * java methods on Android.
 */
const RNFBAnalyticsModule = exports.RNFBAnalyticsModule = {
  logEvent(name, params) {
    return (0, _utils.guard)(async () => {
      const api = getAnalyticsApi('[DEFAULT]');
      api.logEvent(name, params);
      return null;
    });
  },
  setUserId(userId) {
    return (0, _utils.guard)(async () => {
      const api = getAnalyticsApi('[DEFAULT]');
      api.setUserId(userId);
      return null;
    });
  },
  setUserProperty(key, value) {
    return (0, _utils.guard)(async () => {
      const api = getAnalyticsApi('[DEFAULT]');
      api.setUserProperty(key, value);
      return null;
    });
  },
  setUserProperties(properties) {
    return (0, _utils.guard)(async () => {
      const api = getAnalyticsApi('[DEFAULT]');
      api.setUserProperties(properties);
      return null;
    });
  },
  setDefaultEventParameters(params) {
    return (0, _utils.guard)(async () => {
      const api = getAnalyticsApi('[DEFAULT]');
      api.setDefaultEventParameters(params);
      return null;
    });
  },
  setConsent(consent) {
    return (0, _utils.guard)(async () => {
      const api = getAnalyticsApi('[DEFAULT]');
      // TODO currently we only support ad_personalization
      if (consent && consent.ad_personalization !== undefined) {
        api.setConsent({
          ad_personalization: consent.ad_personalization
        });
      }
      return null;
    });
  },
  setAnalyticsCollectionEnabled(enabled) {
    return (0, _utils.guard)(async () => {
      const api = getAnalyticsApi('[DEFAULT]');
      api.setAnalyticsCollectionEnabled(enabled);
      return null;
    });
  },
  resetAnalyticsData() {
    // Unsupported for web.
    return (0, _utils.guard)(async () => {
      return null;
    });
  },
  setSessionTimeoutDuration() {
    // Unsupported for web.
    return (0, _utils.guard)(async () => {
      return null;
    });
  },
  getAppInstanceId() {
    return (0, _utils.guard)(async () => {
      const api = getAnalyticsApi('[DEFAULT]');
      return api._getCid();
    });
  },
  getSessionId() {
    // Unsupported for web.
    return (0, _utils.guard)(async () => {
      return null;
    });
  }
};
