// @react-native-firebase/analytics/dist/commonjs/modular.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getAnalytics = getAnalytics;
exports.getAppInstanceId = getAppInstanceId;
exports.getGoogleAnalyticsClientId = getGoogleAnalyticsClientId;
exports.getSessionId = getSessionId;
exports.initializeAnalytics = initializeAnalytics;
exports.initiateOnDeviceConversionMeasurementWithEmailAddress = initiateOnDeviceConversionMeasurementWithEmailAddress;
exports.initiateOnDeviceConversionMeasurementWithHashedEmailAddress = initiateOnDeviceConversionMeasurementWithHashedEmailAddress;
exports.initiateOnDeviceConversionMeasurementWithHashedPhoneNumber = initiateOnDeviceConversionMeasurementWithHashedPhoneNumber;
exports.initiateOnDeviceConversionMeasurementWithPhoneNumber = initiateOnDeviceConversionMeasurementWithPhoneNumber;
exports.isSupported = isSupported;
exports.logAddPaymentInfo = logAddPaymentInfo;
exports.logAddShippingInfo = logAddShippingInfo;
exports.logAddToCart = logAddToCart;
exports.logAddToWishlist = logAddToWishlist;
exports.logAppOpen = logAppOpen;
exports.logBeginCheckout = logBeginCheckout;
exports.logCampaignDetails = logCampaignDetails;
exports.logEarnVirtualCurrency = logEarnVirtualCurrency;
exports.logEvent = logEvent;
exports.logGenerateLead = logGenerateLead;
exports.logJoinGroup = logJoinGroup;
exports.logLevelEnd = logLevelEnd;
exports.logLevelStart = logLevelStart;
exports.logLevelUp = logLevelUp;
exports.logLogin = logLogin;
exports.logPostScore = logPostScore;
exports.logPurchase = logPurchase;
exports.logRefund = logRefund;
exports.logRemoveFromCart = logRemoveFromCart;
exports.logScreenView = logScreenView;
exports.logSearch = logSearch;
exports.logSelectContent = logSelectContent;
exports.logSelectItem = logSelectItem;
exports.logSelectPromotion = logSelectPromotion;
exports.logSetCheckoutOption = logSetCheckoutOption;
exports.logShare = logShare;
exports.logSignUp = logSignUp;
exports.logSpendVirtualCurrency = logSpendVirtualCurrency;
exports.logTutorialBegin = logTutorialBegin;
exports.logTutorialComplete = logTutorialComplete;
exports.logUnlockAchievement = logUnlockAchievement;
exports.logViewCart = logViewCart;
exports.logViewItem = logViewItem;
exports.logViewItemList = logViewItemList;
exports.logViewPromotion = logViewPromotion;
exports.logViewSearchResults = logViewSearchResults;
exports.resetAnalyticsData = resetAnalyticsData;
exports.setAnalyticsCollectionEnabled = setAnalyticsCollectionEnabled;
exports.setConsent = setConsent;
exports.setDefaultEventParameters = setDefaultEventParameters;
exports.setSessionTimeoutDuration = setSessionTimeoutDuration;
exports.setUserId = setUserId;
exports.setUserProperties = setUserProperties;
exports.setUserProperty = setUserProperty;
exports.settings = settings;
var _common = require("@react-native-firebase/app/lib/common");
var _app = require("@react-native-firebase/app");
var _reactNative = require("react-native");
/**
 * Returns an Analytics instance for the given app.
 */
function getAnalytics(app) {
  if (app) {
    return (0, _app.getApp)(app.name).analytics();
  }
  return (0, _app.getApp)().analytics();
}

/**
 * Returns an Analytics instance for the given app.
 */
function initializeAnalytics(app, _options) {
  return (0, _app.getApp)(app.name).analytics();
}

/**
 * Retrieves a unique Google Analytics identifier for the web client.
 */
async function getGoogleAnalyticsClientId(analytics) {
  if (_reactNative.Platform.OS === 'android' || _reactNative.Platform.OS === 'ios') {
    throw new Error('getGoogleAnalyticsClientId is web-only.');
  } else {
    const instanceId = await getAppInstanceId(analytics);
    return instanceId || '';
  }
}

/**
 * Log a custom event with optional params.
 */
// Overloads for standard event names

// Generic fallback for custom event names

function logEvent(analytics, name, params = {}, options = {
  global: false
}) {
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.logEvent.call(analytics, name, params, options, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * If true, allows the device to collect analytical data and send it to Firebase. Useful for GDPR.
 */
function setAnalyticsCollectionEnabled(analytics, enabled) {
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.setAnalyticsCollectionEnabled.call(analytics, enabled, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * Sets the duration of inactivity that terminates the current session.
 */
function setSessionTimeoutDuration(analytics, milliseconds = 1800000) {
  // This doesn't exist on firebase-js-sdk, but probably should keep this for android & iOS
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.setSessionTimeoutDuration.call(analytics, milliseconds, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * Retrieve the app instance id of the application.
 */
function getAppInstanceId(analytics) {
  // This doesn't exist on firebase-js-sdk, but probably should keep this for android & iOS
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.getAppInstanceId.call(analytics, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * Retrieves the session id from the client.
 * On iOS, Firebase SDK may return an error that is handled internally and may take many minutes to return a valid value. Check native debug logs for more details.
 */
function getSessionId(analytics) {
  // This doesn't exist on firebase-js-sdk, but probably should keep this for android & iOS
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.getSessionId.call(analytics, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * Gives a user a unique identification.
 */
function setUserId(analytics, id) {
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.setUserId.call(analytics, id, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * Sets a key/value pair of data on the current user. Each Firebase project can have up to 25 uniquely named (case-sensitive) user properties.
 */
function setUserProperty(analytics, name, value) {
  // This doesn't exist on firebase-js-sdk, but probably should keep this for android & iOS
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.setUserProperty.call(analytics, name, value, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * Sets multiple key/value pairs of data on the current user. Each Firebase project can have up to 25 uniquely named (case-sensitive) user properties.
 */
function setUserProperties(analytics, properties, options = {
  global: false
}) {
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.setUserProperties.call(analytics, properties, options, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * Clears all analytics data for this instance from the device and resets the app instance ID.
 */
function resetAnalyticsData(analytics) {
  // This doesn't exist on firebase-js-sdk, but probably should keep this for android & iOS
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.resetAnalyticsData.call(analytics, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * E-Commerce Purchase event. This event signifies that an item(s) was purchased by a user. Note: This is different from the in-app purchase event, which is reported automatically for Google Play-based apps.
 */
function logAddPaymentInfo(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logAddPaymentInfo(params);
}

/**
 * Sets or clears the screen name and class the user is currently viewing.
 */
function logScreenView(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logScreenView(params);
}

/**
 * Add Payment Info event. This event signifies that a user has submitted their payment information to your app.
 */
function logAddShippingInfo(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logAddShippingInfo(params);
}

/**
 * E-Commerce Add To Cart event.
 */
function logAddToCart(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logAddToCart(params);
}

/**
 * E-Commerce Add To Wishlist event. This event signifies that an item was added to a wishlist.
 */
function logAddToWishlist(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logAddToWishlist(params);
}

/**
 * App Open event. By logging this event when an App is moved to the foreground, developers can understand how often users leave and return during the course of a Session.
 */
function logAppOpen(analytics) {
  // This is deprecated for both namespaced and modular.
  return analytics.logAppOpen();
}

/**
 * E-Commerce Begin Checkout event. This event signifies that a user has begun the process of checking out.
 */
function logBeginCheckout(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logBeginCheckout(params);
}

/**
 * Log this event to supply the referral details of a re-engagement campaign.
 */
function logCampaignDetails(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logCampaignDetails(params);
}

/**
 * Earn Virtual Currency event. This event tracks the awarding of virtual currency in your app.
 */
function logEarnVirtualCurrency(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logEarnVirtualCurrency(params);
}

/**
 * Generate Lead event. Log this event when a lead has been generated in the app.
 */
function logGenerateLead(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logGenerateLead(params);
}

/**
 * Join Group event. Log this event when a user joins a group such as a guild, team or family.
 */
function logJoinGroup(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logJoinGroup(params);
}

/**
 * Level End event.
 */
function logLevelEnd(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logLevelEnd(params);
}

/**
 * Level Start event.
 */
function logLevelStart(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logLevelStart(params);
}

/**
 * Level Up event. This event signifies that a player has leveled up in your gaming app.
 */
function logLevelUp(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logLevelUp(params);
}

/**
 * Login event. Apps with a login feature can report this event to signify that a user has logged in.
 */
function logLogin(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logLogin(params);
}

/**
 * Post Score event. Log this event when the user posts a score in your gaming app.
 */
function logPostScore(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logPostScore(params);
}

/**
 * Select Content event. This general purpose event signifies that a user has selected some content of a certain type in an app.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {SelectContentEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logSelectContent(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logSelectContent(params);
}

/**
 * E-Commerce Purchase event. This event signifies that an item(s) was purchased by a user.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {PurchaseEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logPurchase(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logPurchase(params);
}

/**
 * E-Commerce Refund event. This event signifies that a refund was issued.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {RefundEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logRefund(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logRefund(params);
}

/**
 * Remove from cart event.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {RemoveFromCartEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logRemoveFromCart(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logRemoveFromCart(params);
}

/**
 * Search event. Apps that support search features can use this event to contextualize search operations by supplying the appropriate, corresponding parameters.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {SearchEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logSearch(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logSearch(params);
}

/**
 * Select Item event. This event signifies that an item was selected by a user from a list.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {SelectItemEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logSelectItem(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logSelectItem(params);
}

/**
 * Set checkout option event.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {SetCheckoutOptionEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logSetCheckoutOption(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logSetCheckoutOption(params);
}

/**
 * Select promotion event. This event signifies that a user has selected a promotion offer.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {SelectPromotionEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logSelectPromotion(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logSelectPromotion(params);
}

/**
 * Share event. Apps with social features can log the Share event to identify the most viral content.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {ShareEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logShare(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logShare(params);
}

/**
 * Sign Up event. This event indicates that a user has signed up for an account in your app.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {SignUpEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logSignUp(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logSignUp(params);
}

/**
 * Spend Virtual Currency event. This event tracks the sale of virtual goods in your app.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {SpendVirtualCurrencyEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logSpendVirtualCurrency(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logSpendVirtualCurrency(params);
}

/**
 * Tutorial Begin event. This event signifies the start of the on-boarding process in your app.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @returns {Promise<void>}
 */
function logTutorialBegin(analytics) {
  // This is deprecated for both namespaced and modular.
  return analytics.logTutorialBegin();
}

/**
 * Tutorial End event. Use this event to signify the user's completion of your app's on-boarding process.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @returns {Promise<void>}
 */
function logTutorialComplete(analytics) {
  // This is deprecated for both namespaced and modular.
  return analytics.logTutorialComplete();
}

/**
 * Unlock Achievement event. Log this event when the user has unlocked an achievement in your game.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {UnlockAchievementEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logUnlockAchievement(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logUnlockAchievement(params);
}

/**
 * E-commerce View Cart event. This event signifies that a user has viewed their cart.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {ViewCartEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logViewCart(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logViewCart(params);
}

/**
 * View Item event. This event signifies that some content was shown to the user.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {ViewItemEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logViewItem(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logViewItem(params);
}

/**
 * View Item List event. Log this event when the user has been presented with a list of items of a certain category.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {ViewItemListEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logViewItemList(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logViewItemList(params);
}

/**
 * View Promotion event. This event signifies that a promotion was shown to a user.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {ViewPromotionEventParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logViewPromotion(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logViewPromotion(params);
}

/**
 * View Search Results event. Log this event when the user has been presented with the results of a search.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {ViewSearchResultsParameters} params - Event parameters.
 * @returns {Promise<void>}
 */
function logViewSearchResults(analytics, params) {
  // This is deprecated for both namespaced and modular.
  return analytics.logViewSearchResults(params);
}

/**
 * Adds parameters that will be set on every event logged from the SDK, including automatic ones.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {object} [params={}] - Parameters to be added to the map of parameters added to every event.
 * @returns {Promise<void>}
 */
function setDefaultEventParameters(analytics, params = {}) {
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.setDefaultEventParameters.call(analytics, params, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * start privacy-sensitive on-device conversion management.
 * This is iOS-only.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {string} emailAddress - Email address, properly formatted complete with domain name e.g, 'user@example.com'.
 * @returns {Promise<void>}
 */
function initiateOnDeviceConversionMeasurementWithEmailAddress(analytics, emailAddress) {
  return analytics.initiateOnDeviceConversionMeasurementWithEmailAddress.call(analytics, emailAddress,
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  _common.MODULAR_DEPRECATION_ARG);
}

/**
 * start privacy-sensitive on-device conversion management.
 * This is iOS-only.
 * This is a no-op if you do not include '$RNFirebaseAnalyticsGoogleAppMeasurementOnDeviceConversion = true' in your Podfile
 *
 * @param analytics Analytics instance.
 * @param hashedEmailAddress sha256-hashed of normalized email address, properly formatted complete with domain name e.g, 'user@example.com'
 * @link https://firebase.google.com/docs/tutorials/ads-ios-on-device-measurement/step-3#use-hashed-credentials
 */
function initiateOnDeviceConversionMeasurementWithHashedEmailAddress(analytics, hashedEmailAddress) {
  return analytics.initiateOnDeviceConversionMeasurementWithHashedEmailAddress.call(analytics, hashedEmailAddress,
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  _common.MODULAR_DEPRECATION_ARG);
}

/**
 * start privacy-sensitive on-device conversion management.
 * This is iOS-only.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {string} phoneNumber - Phone number in E.164 format - that is a leading + sign, then up to 15 digits, no dashes or spaces.
 * @returns {Promise<void>}
 */
function initiateOnDeviceConversionMeasurementWithPhoneNumber(analytics, phoneNumber) {
  return analytics.initiateOnDeviceConversionMeasurementWithPhoneNumber.call(analytics, phoneNumber,
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  _common.MODULAR_DEPRECATION_ARG);
}

/**
 * start privacy-sensitive on-device conversion management.
 * This is iOS-only.
 * This is a no-op if you do not include '$RNFirebaseAnalyticsGoogleAppMeasurementOnDeviceConversion = true' in your Podfile
 *
 * @param analytics Analytics instance.
 * @param hashedPhoneNumber sha256-hashed of normalized phone number in E.164 format - that is a leading + sign, then up to 15 digits, no dashes or spaces.
 * @link https://firebase.google.com/docs/tutorials/ads-ios-on-device-measurement/step-3#use-hashed-credentials
 */
function initiateOnDeviceConversionMeasurementWithHashedPhoneNumber(analytics, hashedPhoneNumber) {
  return analytics.initiateOnDeviceConversionMeasurementWithHashedPhoneNumber.call(analytics, hashedPhoneNumber,
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  _common.MODULAR_DEPRECATION_ARG);
}

/**
 * Checks four different things.
 * 1. Checks if it's not a browser extension environment.
 * 2. Checks if cookies are enabled in current browser.
 * 3. Checks if IndexedDB is supported by the browser environment.
 * 4. Checks if the current browser context is valid for using IndexedDB.open().
 * @returns {Promise<boolean>}
 */
function isSupported() {
  return Promise.resolve(true);
}

/**
 * Sets the applicable end user consent state for this app.
 * @param {FirebaseAnalytics} analytics - Analytics instance.
 * @param {ConsentSettings} consentSettings - See ConsentSettings.
 * @returns {Promise<void>}
 */
function setConsent(analytics, consentSettings) {
  // @ts-ignore - MODULAR_DEPRECATION_ARG is filtered out internally
  return analytics.setConsent.call(analytics, consentSettings, _common.MODULAR_DEPRECATION_ARG);
}

/**
 * Configures Firebase Analytics to use custom gtag or dataLayer names.
 * Intended to be used if gtag.js script has been installed on this page independently of Firebase Analytics, and is using non-default names for either the gtag function or for dataLayer. Must be called before calling `getAnalytics()` or it won't have any effect. Web only.
 * @param {SettingsOptions} options - See SettingsOptions.
 * @returns {void}
 */
function settings(_options) {
  // Returns nothing until Web implemented.
}
