// @react-native-firebase/analytics/dist/commonjs/structs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ViewSearchResults = exports.ViewPromotion = exports.ViewItemList = exports.ViewItem = exports.ViewCart = exports.UnlockAchievement = exports.SpendVirtualCurrency = exports.SignUp = exports.Share = exports.SetCheckoutOption = exports.SelectPromotion = exports.SelectItem = exports.SelectContent = exports.Search = exports.ScreenView = exports.RemoveFromCart = exports.Refund = exports.Purchase = exports.PostScore = exports.Login = exports.LevelUp = exports.LevelStart = exports.LevelEnd = exports.JoinGroup = exports.GenerateLead = exports.EarnVirtualCurrency = exports.CampaignDetails = exports.BeginCheckout = exports.AddToWishlist = exports.AddToCart = exports.AddShippingInfo = exports.AddPaymentInfo = void 0;
var _superstruct = require("superstruct");
/*
 *  Copyright (c) 2016-present Invertase Limited & Contributors
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this library except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

const ShortDate = (0, _superstruct.define)('ShortDate', value => typeof value === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(value));
const Item = (0, _superstruct.type)({
  item_brand: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_id: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_name: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_category: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_category2: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_category3: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_category4: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_category5: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_list_id: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_list_name: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_location_id: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_variant: (0, _superstruct.optional)((0, _superstruct.string)()),
  quantity: (0, _superstruct.optional)((0, _superstruct.number)()),
  price: (0, _superstruct.optional)((0, _superstruct.number)())
});
const ScreenView = exports.ScreenView = (0, _superstruct.type)({
  screen_class: (0, _superstruct.optional)((0, _superstruct.string)()),
  screen_name: (0, _superstruct.optional)((0, _superstruct.string)())
});
const AddPaymentInfo = exports.AddPaymentInfo = (0, _superstruct.object)({
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  value: (0, _superstruct.optional)((0, _superstruct.number)()),
  currency: (0, _superstruct.optional)((0, _superstruct.string)()),
  coupon: (0, _superstruct.optional)((0, _superstruct.string)()),
  payment_type: (0, _superstruct.optional)((0, _superstruct.string)())
});
const AddShippingInfo = exports.AddShippingInfo = (0, _superstruct.object)({
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  value: (0, _superstruct.optional)((0, _superstruct.number)()),
  currency: (0, _superstruct.optional)((0, _superstruct.string)()),
  coupon: (0, _superstruct.optional)((0, _superstruct.string)()),
  shipping_tier: (0, _superstruct.optional)((0, _superstruct.string)())
});
const AddToCart = exports.AddToCart = (0, _superstruct.object)({
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  value: (0, _superstruct.optional)((0, _superstruct.number)()),
  currency: (0, _superstruct.optional)((0, _superstruct.string)())
});
const AddToWishlist = exports.AddToWishlist = (0, _superstruct.object)({
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  value: (0, _superstruct.optional)((0, _superstruct.number)()),
  currency: (0, _superstruct.optional)((0, _superstruct.string)())
});
const BeginCheckout = exports.BeginCheckout = (0, _superstruct.type)({
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  value: (0, _superstruct.optional)((0, _superstruct.number)()),
  currency: (0, _superstruct.optional)((0, _superstruct.string)()),
  coupon: (0, _superstruct.optional)((0, _superstruct.string)())
});
const CampaignDetails = exports.CampaignDetails = (0, _superstruct.object)({
  source: (0, _superstruct.string)(),
  medium: (0, _superstruct.string)(),
  campaign: (0, _superstruct.string)(),
  term: (0, _superstruct.optional)((0, _superstruct.string)()),
  content: (0, _superstruct.optional)((0, _superstruct.string)()),
  aclid: (0, _superstruct.optional)((0, _superstruct.string)()),
  cp1: (0, _superstruct.optional)((0, _superstruct.string)())
});
const EarnVirtualCurrency = exports.EarnVirtualCurrency = (0, _superstruct.object)({
  virtual_currency_name: (0, _superstruct.string)(),
  value: (0, _superstruct.number)()
});
const GenerateLead = exports.GenerateLead = (0, _superstruct.object)({
  currency: (0, _superstruct.optional)((0, _superstruct.string)()),
  value: (0, _superstruct.optional)((0, _superstruct.number)())
});
const JoinGroup = exports.JoinGroup = (0, _superstruct.object)({
  group_id: (0, _superstruct.string)()
});
const LevelEnd = exports.LevelEnd = (0, _superstruct.object)({
  level: (0, _superstruct.number)(),
  success: (0, _superstruct.optional)((0, _superstruct.string)())
});
const LevelStart = exports.LevelStart = (0, _superstruct.object)({
  level: (0, _superstruct.number)()
});
const LevelUp = exports.LevelUp = (0, _superstruct.object)({
  level: (0, _superstruct.number)(),
  character: (0, _superstruct.optional)((0, _superstruct.string)())
});
const Login = exports.Login = (0, _superstruct.object)({
  method: (0, _superstruct.string)()
});
const PostScore = exports.PostScore = (0, _superstruct.object)({
  score: (0, _superstruct.number)(),
  level: (0, _superstruct.optional)((0, _superstruct.number)()),
  character: (0, _superstruct.optional)((0, _superstruct.string)())
});
const Refund = exports.Refund = (0, _superstruct.object)({
  affiliation: (0, _superstruct.optional)((0, _superstruct.string)()),
  coupon: (0, _superstruct.optional)((0, _superstruct.string)()),
  currency: (0, _superstruct.optional)((0, _superstruct.string)()),
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  shipping: (0, _superstruct.optional)((0, _superstruct.number)()),
  tax: (0, _superstruct.optional)((0, _superstruct.number)()),
  value: (0, _superstruct.optional)((0, _superstruct.number)()),
  transaction_id: (0, _superstruct.optional)((0, _superstruct.string)())
});
const Purchase = exports.Purchase = (0, _superstruct.type)({
  affiliation: (0, _superstruct.optional)((0, _superstruct.string)()),
  coupon: (0, _superstruct.optional)((0, _superstruct.string)()),
  currency: (0, _superstruct.optional)((0, _superstruct.string)()),
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  shipping: (0, _superstruct.optional)((0, _superstruct.number)()),
  tax: (0, _superstruct.optional)((0, _superstruct.number)()),
  value: (0, _superstruct.optional)((0, _superstruct.number)()),
  transaction_id: (0, _superstruct.optional)((0, _superstruct.string)())
});
const RemoveFromCart = exports.RemoveFromCart = (0, _superstruct.object)({
  currency: (0, _superstruct.optional)((0, _superstruct.string)()),
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  value: (0, _superstruct.optional)((0, _superstruct.number)())
});
const Search = exports.Search = (0, _superstruct.object)({
  search_term: (0, _superstruct.string)(),
  number_of_nights: (0, _superstruct.optional)((0, _superstruct.number)()),
  number_of_rooms: (0, _superstruct.optional)((0, _superstruct.number)()),
  number_of_passengers: (0, _superstruct.optional)((0, _superstruct.number)()),
  origin: (0, _superstruct.optional)((0, _superstruct.string)()),
  destination: (0, _superstruct.optional)((0, _superstruct.string)()),
  start_date: (0, _superstruct.optional)(ShortDate),
  end_date: (0, _superstruct.optional)(ShortDate),
  travel_class: (0, _superstruct.optional)((0, _superstruct.string)())
});
const SelectContent = exports.SelectContent = (0, _superstruct.object)({
  content_type: (0, _superstruct.string)(),
  item_id: (0, _superstruct.string)()
});
const SelectItem = exports.SelectItem = (0, _superstruct.object)({
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  item_list_id: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_list_name: (0, _superstruct.optional)((0, _superstruct.string)()),
  content_type: (0, _superstruct.optional)((0, _superstruct.string)())
});
const SelectPromotion = exports.SelectPromotion = (0, _superstruct.object)({
  creative_name: (0, _superstruct.string)(),
  creative_slot: (0, _superstruct.string)(),
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  location_id: (0, _superstruct.string)(),
  promotion_id: (0, _superstruct.string)(),
  promotion_name: (0, _superstruct.string)()
});
const SetCheckoutOption = exports.SetCheckoutOption = (0, _superstruct.type)({
  checkout_step: (0, _superstruct.optional)((0, _superstruct.number)()),
  checkout_option: (0, _superstruct.optional)((0, _superstruct.string)())
});
const Share = exports.Share = (0, _superstruct.object)({
  content_type: (0, _superstruct.string)(),
  item_id: (0, _superstruct.string)(),
  method: (0, _superstruct.string)()
});
const SignUp = exports.SignUp = (0, _superstruct.object)({
  method: (0, _superstruct.string)()
});
const SpendVirtualCurrency = exports.SpendVirtualCurrency = (0, _superstruct.object)({
  item_name: (0, _superstruct.string)(),
  virtual_currency_name: (0, _superstruct.string)(),
  value: (0, _superstruct.number)()
});
const UnlockAchievement = exports.UnlockAchievement = (0, _superstruct.object)({
  achievement_id: (0, _superstruct.string)()
});
const ViewCart = exports.ViewCart = (0, _superstruct.object)({
  currency: (0, _superstruct.optional)((0, _superstruct.string)()),
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  value: (0, _superstruct.optional)((0, _superstruct.number)())
});
const ViewItem = exports.ViewItem = (0, _superstruct.object)({
  currency: (0, _superstruct.optional)((0, _superstruct.string)()),
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  value: (0, _superstruct.optional)((0, _superstruct.number)())
});
const ViewItemList = exports.ViewItemList = (0, _superstruct.object)({
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  item_list_id: (0, _superstruct.optional)((0, _superstruct.string)()),
  item_list_name: (0, _superstruct.optional)((0, _superstruct.string)())
});
const ViewPromotion = exports.ViewPromotion = (0, _superstruct.object)({
  items: (0, _superstruct.optional)((0, _superstruct.array)(Item)),
  location_id: (0, _superstruct.optional)((0, _superstruct.string)()),
  creative_name: (0, _superstruct.optional)((0, _superstruct.string)()),
  creative_slot: (0, _superstruct.optional)((0, _superstruct.string)()),
  promotion_id: (0, _superstruct.optional)((0, _superstruct.string)()),
  promotion_name: (0, _superstruct.optional)((0, _superstruct.string)())
});
const ViewSearchResults = exports.ViewSearchResults = (0, _superstruct.object)({
  search_term: (0, _superstruct.string)()
});
