// @react-native-firebase/app/dist/commonjs/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _index = require("../common/index.js");
var _index2 = require("../internal/index.js");
var _UtilsStatics = _interopRequireDefault(require("./UtilsStatics.js"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

const namespace = 'utils';
const statics = _UtilsStatics.default;
const nativeModuleName = 'RNFBUtilsModule';
class FirebaseUtilsModule extends _index2.FirebaseModule {
  get isRunningInTestLab() {
    if (_index.isIOS) {
      return false;
    }
    return this.native.isRunningInTestLab;
  }
  get playServicesAvailability() {
    if (_index.isIOS) {
      return {
        isAvailable: true,
        status: 0,
        hasResolution: false,
        isUserResolvableError: false,
        error: undefined
      };
    }
    return this.native.androidPlayServices;
  }
  getPlayServicesStatus() {
    if (_index.isIOS) {
      return Promise.resolve({
        isAvailable: true,
        status: 0,
        hasResolution: false,
        isUserResolvableError: false,
        error: undefined
      });
    }
    return this.native.androidGetPlayServicesStatus();
  }
  promptForPlayServices() {
    if (_index.isIOS) {
      return Promise.resolve();
    }
    return this.native.androidPromptForPlayServices();
  }
  makePlayServicesAvailable() {
    if (_index.isIOS) {
      return Promise.resolve();
    }
    return this.native.androidMakePlayServicesAvailable();
  }
  resolutionForPlayServices() {
    if (_index.isIOS) {
      return Promise.resolve();
    }
    return this.native.androidResolutionForPlayServices();
  }
}

// import { utils } from '@react-native-firebase/app';
// utils().X(...);
var _default = exports.default = (0, _index2.createModuleNamespace)({
  statics,
  version: _UtilsStatics.default.SDK_VERSION,
  namespace,
  nativeModuleName,
  nativeEvents: false,
  hasMultiAppSupport: false,
  hasCustomUrlOrRegionSupport: false,
  ModuleClass: FirebaseUtilsModule
});
